### PR TITLE
Function Adds some properties

### PR DESCRIPTION
**Pull request by hedaoyuan**
Add FunctionBase::ops() and FunctionBase::check()

实现ISSUE中的特性
https://github.com/PaddlePaddle/Paddle/issues/892#issuecomment-267358563

### CODE REVIEW

File: paddle/function/CrossMapNormalOp.cpp [hunk 1]
```diff
@@ -162,38 +162,64 @@ template <DeviceType Device>
...
-    CHECK_EQ((size_t)1, inputs.size());
-    CHECK_EQ((size_t)2, outputs.size());
-
-    CHECK_EQ(inputs[0].shape().ndims(), (size_t)4);
-    CHECK(inputs[0].shape() == outputs[0].shape());
-    CHECK(inputs[0].shape() == outputs[1].shape());
-
+    check(inputs, outputs);
+    // ArgType check still on here,
+    // not sure whether it is better to put inside the check.
```
Comment (tianbingsz) [inline paddle/function/CrossMapNormalOp.cpp]: I think we can move the CHECK of argType into check funciton.

File: paddle/function/Function.h [hunk 2]
```diff
@@ -153,7 +153,36 @@ class FunctionBase {
...
+  // And before the calc called, the caller can also check their own arguments.
+  virtual void check(const BufferArgs& inputs, const BufferArgs& outputs) {}
+
+  // Calculate the number of floating-point operations of this Function.
+  // The inputs and outputs arguments do not need to contain the actual data,
+  // only the shape.
+  // And some Functions have the same input and output shapes,
+  // so you may not need to enter the complete number of arguments.
+  // But entering the full arguments is always correct for this interface.
+  virtual size_t ops(const BufferArgs& inputs, const BufferArgs& outputs) {
```
Comment (tianbingsz) [inline paddle/function/Function.h]: Can we use timestamp to measure the running time of the operators?

Comment (hedaoyuan) [inline paddle/function/Function.h]: 嗯，后续再考虑如何获取运行实现，这里先描述operators数。

File: paddle/function/CrossMapNormalOp.cpp (hunk 1 above)
Comment (hedaoyuan) [inline paddle/function/CrossMapNormalOp.cpp]: 我觉得先放calc里面吧，API的参数可能跟ASSIGN_TO/ADD_TO有关，放这里会清楚一些。

**Pull request merged**